src/schema/posts.ts

```diff
@@ -1516,6 +1635,112 @@ export const resolvers: IResolvers<any, Context> = {
       });
       return { _: true };
     },
+    votePost: async (
+      source,
+      { id, vote }: { id: string; vote: UserPostVote },
+      ctx: Context,
+    ): Promise<GQLEmptyResponse> => {
+      try {
+        if (!Object.values(UserPostVote).includes(vote)) {
+          throw new ValidationError('Unsupported vote type');
+        }
+
+        const post = await ctx.con.getRepository(Post).findOneByOrFail({ id });
+        await ensureSourcePermissions(ctx, post.sourceId);
+        const userPostRepo = ctx.con.getRepository(UserPost);
+
+        switch (vote) {
+          case UserPostVote.Up:
+            await userPostRepo.save({
+              postId: id,
+              userId: ctx.userId,
+              vote: UserPostVote.Up,
+              hidden: false,
+            });
+
+            break;
+          case UserPostVote.Down:
+            await userPostRepo.save({
+              postId: id,
+              userId: ctx.userId,
+              vote: UserPostVote.Down,
+              hidden: true,
+            });
+
+            break;
+          case UserPostVote.None:
+            await userPostRepo.save({
+              postId: id,
+              userId: ctx.userId,
+              vote: UserPostVote.None,
+              hidden: false,
+            });
+
+            break;
+          default:
+            throw new ValidationError('Unsupported vote type');
+        }
+      } catch (err) {
+        // Foreign key violation
+        if (err?.code === TypeOrmError.FOREIGN_KEY) {
+          throw new NotFoundError('Post or user not found');
+        }
+
+        throw err;
+      }
+
+      return { _: true };
+    },
+    dismissPostFeedback: async (
+      source,
+      { id }: { id: string },
+      ctx: Context,
+    ) => {
+      try {
+        const post = await ctx.con.getRepository(Post).findOneByOrFail({ id });
+        await ensureSourcePermissions(ctx, post.sourceId);
+
+        await ctx.con.transaction(async (entityManager) => {
+          const userPost = await entityManager
+            .getRepository(UserPost)
+            .findOneBy({
+              postId: id,
+              userId: ctx.userId,
+            });
+
+          if (userPost) {
+            await entityManager.getRepository(UserPost).update(
+              {
+                postId: id,
+                userId: ctx.userId,
+              },
+              {
+                flags: updateFlagsStatement<UserPost>({
+                  feedbackDismiss: true,
+                }),
+              },
+            );
+          } else {
+            await entityManager.getRepository(UserPost).insert({
+              postId: id,
+              userId: ctx.userId,
+              flags: {
+                feedbackDismiss: true,
+              },
+            });
+          }
```

Comment:
Had to do this since `save` does not support custom update queries and we would overwrite other flags in the future if just doing `save` since `typeorm` does not support nested merging of json fields. That's why we needed to have `updateFlagsStatement` helper.